### ISSUE TITLE
Communication between Reliable stateful actor client and service taking long time

### ISSUE THREAD

**Issue by prinupmathew** [Azure/service-fabric-issues#316]
Please help me to understand what is the cause of this delay

Runtime Version: 5.6.220.9494
SDK: 2.6.204

I am experiencing long delay (over 60 seconds) when initial communication is established between Reliable actor client proxy and service. This works fine when deployed and tested in local cluster. 
But start having communication delay when deployed in remote cluster (in my case AWS cloud). This delay occurs again after certain numbers of calls are made successfully. Here is the pattern that i see
w.r.t to request-response time

Actor proxy is executed from Reliable service to Reliable actor

Below proxy calls are in iterations

First Iteration

Initial method call - over 60 secs
2nd call - less than 200 ms,
3rd call - less than 200 ms,
4th call - less than 200 ms,
5th call - less than 200 ms,

Second Iteration

Initial method call - over 70 secs,
2nd call - less than 200 ms,
3rd call - less than 200 ms,
4th call - less than 200 ms,
5th call - less than 200 ms

Third Iteration

Initial method call - over 70 secs,
2nd call - less than 200 ms,
3rd call - less than 200 ms,
4th call - more than 25 secs,
5th call - less than 200 ms

Here is the code snippet

**Reliable Actor Service Implementations**
  
protected override IEnumerable<ServiceReplicaListener> CreateServiceReplicaListeners()
        {
            return new[]
            {
                new ServiceReplicaListener((context) => new FabricTransportActorServiceRemotingListener(this, new FabricTransportRemotingListenerSettings
                {
                    EndpointResourceName = "ScheduleWorkerActorServiceEndpoint"
                }), "RemotingServiceEndpointListener")
            };
        }
		
**ServiceManifest.xml**
		
```
<?xml version="1.0" encoding="utf-8"?>
<ServiceManifest xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Name="ScheduleActorPkg" Version="1.0.0" xmlns="http://schemas.microsoft.com/2011/01/fabric">
  <ServiceTypes>   
    <StatefulServiceType ServiceTypeName="ScheduleWorkerActorServiceType" HasPersistedState="true">
      <Extensions>
        <Extension Name="__GeneratedServiceType__" GeneratedId="dfc824d9-5788-4c44-85a1-d9f47f7b6b37|Persisted">
          <GeneratedNames xmlns="http://schemas.microsoft.com/2015/03/fabact-no-schema">
            <DefaultService Name="ScheduleWorkerActorService" />
            <ServiceEndpoint Name="ScheduleWorkerActorServiceEndpoint" />
            <ReplicatorEndpoint Name="ScheduleWorkerActorServiceReplicatorEndpoint" />
            <ReplicatorConfigSection Name="ScheduleWorkerActorServiceReplicatorConfig" />
            <ReplicatorSecurityConfigSection Name="ScheduleWorkerActorServiceReplicatorSecurityConfig" />
            <StoreConfigSection Name="ScheduleWorkerActorServiceLocalStoreConfig" />
          </GeneratedNames>
        </Extension>
      </Extensions>
    </StatefulServiceType>
  </ServiceTypes>
 <CodePackage Name="Code" Version="1.0.0">
    <EntryPoint>
      <ExeHost>
        <Program>ScheduleActor.exe</Program>
      </ExeHost>
    </EntryPoint>
  </CodePackage>
<ConfigPackage Name="Config" Version="1.0.0" />
  <Resources>
    <Endpoints>
      <Endpoint Name="ScheduleWorkerActorServiceEndpoint" />
      <Endpoint Name="ScheduleWorkerActorServiceReplicatorEndpoint" />    
    </Endpoints>
  </Resources>
  <!-- The content will be generated during build -->
</ServiceManifest>
```

**Settings.xml**

```
<?xml version="1.0" encoding="utf-8"?>
<Settings xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://schemas.microsoft.com/2011/01/fabric">
  <Section Name="ScheduleWorkerActorServiceReplicatorConfig">
    <Parameter Name="ReplicatorEndpoint" Value="ScheduleWorkerActorServiceReplicatorEndpoint" />
    <Parameter Name="BatchAcknowledgementInterval" Value="0.005" />
  </Section>
  <Section Name="ScheduleWorkerActorServiceReplicatorSecurityConfig">
    <Parameter Name="CredentialType" Value="None" />
  </Section>  
  <Section Name="TransportSettings">
    <Parameter Name="MaxMessageSize" Value="134217728" />
    <Parameter Name="OperationTimeoutInSeconds" Value="300" />
    <Parameter Name="MaxConcurrentCalls" Value="64" />
  </Section>
</Settings>
```

**Reliable Actor Client Implementations**

var random = new Random();
var workerId = $"worker_{random.Next(1, 101)}";

ActorProxyFactory proxyFactory = new ActorProxyFactory();
          
var scheduleWorkerActoProxy = proxyFactory.CreateActorProxy<IScheduleWorkerActor>(ScheduleWorkerActorServiceUri, new ActorId(workerId), listenerName: "RemotingServiceEndpointListener");

**Settings.xml**

```
<?xml version="1.0" encoding="utf-8" ?>
<Settings xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://schemas.microsoft.com/2011/01/fabric">
  <Section Name="TransportSettings">
    <Parameter Name="MaxMessageSize" Value="134217728" />
    <Parameter Name="OperationTimeoutInSeconds" Value="300" />
    <Parameter Name="MaxConcurrentCalls" Value="64" />
  </Section>
</Settings>
```

**Comment by suchiagicha** [comment]
@prinupmathew 

Could you provide more insight on how are you measuring time?
Does it includes proxyFactory.CreateActorProxy  execution time or its only method call on actorProxy or both?
Are you using creating actorProxy for each iteration ?
Is it possible for you to share to the repro code for your actor Service and Client?

**Comment by prinupmathew** [comment]
@suchiagicha 

I have added stop watch to measure the request and response between actor client and actor service

Attached the source code which contains 5 projects taken from a POC solution. Below summarize the critical area where the delay occurs. Please let me know if you need more information.

In project ScheduleService, 

class: ScheduleService
method: GetXtraSchedulesParallel
line: 206

Below line of code calls the Actor service (inside a foreach loop of 30 iterations) which taking more than 60 secs to initialize the communication channel. This delay is just establishing the connection between actor client and service. Once the connection is established then the actor response without issue. Again this connection delay occurs after few iterations. Unfortunately i couldn't find any useful error logs or am i missing something? 
`  
//Start parallel processing
                if (!await scheduleWorkerActoProxy.StartParallelProcessingAsync(getXtraScheduleRequestModelTask))
                {
                    continue;
                }`

[Actor Delay.zip](https://github.com/Azure/service-fabric-issues/files/1079751/Actor.Delay.zip)

**Comment by suchiagicha** [comment]
@prinupmathew 
Few things I noticed at first glance.
This is not a working repro . I can't run on my box to confirm your theory.
Reviewing your code :
You are creating scheduleWorkerActoProxy and then invoke 3 Api for it. 
1 All 3 Api will use the same communication Channel for same proxy if there is no failover for your Actor Service. Hence delay in communication is less likely.

2 SubscribeAsync , IsParallelProcessingRunningAsync and StartParallelProcessingAsync are 3 Api .
You are seeing high latency only for 3rd  Api call (StartParallelProcessingAsync).

There can be 2 possibilities if that is the case :
 1  Your StartParallelProcessingAsync Implmentation has some logic which will delay it for first time. Check your implementation . Have some instrumentation around it.
2  If any method of an Actor is being processed around the same time, your other API calls for the same actor  has to wait for it to complete. In Actor  no more than one thread can be active inside an actor object's code at any time. 
https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-actors-introduction

**Comment by prinupmathew** [comment]
@suchiagicha 

The time actor calls StartParallelProcessingAsync method is taking time and the logic inside that method is not the one causing the issue. Once the actor hits the method then processing of the method is faster.

What is don't understand is why actor runtime takes time to establish connection between  client and service.

If possible then is there any way i can share my screen and show the demo?

**Comment by suchiagicha** [comment]
Also you should not be creating   ActorProxyFactory  for each call GetXtraSchedulesParallel. It is an expensive operation. You should try to re-use it , this will have a huge performance impact if they are multiple calls for GetXtraSchedulesParallel .
Or use the default proxy factory , if you are not overriding the factory settings.
like this ActorProxy.Create<>();

**Comment by prinupmathew** [comment]
I understand that ActorProxyFactory is expensive calls and should be initialized only once. 

As per current POC, GetXtraSchedulesParallel is called only once per client. The real bottle neck is when below code is executed in an iteration

```
  ActorProxyFactory proxyFactory = new ActorProxyFactory();

                var scheduleWorkerActoProxy = proxyFactory.CreateActorProxy<IScheduleWorkerActor>(ScheduleWorkerActorServiceUri, workerActorId, listenerName: "RemotingServiceEndpointListener");

 //Start parallel processing
                if (!await scheduleWorkerActoProxy.StartParallelProcessingAsync(xtraScheduleTask))
                {
                    continue;
                }
```

**Comment by prinupmathew** [comment]
@suchiagicha 

As per our discussion, attached the latest 2 trace logs from AWS cluster 3 nodes


[fabric_traces_5.6.220.9494_131412591192127891_339.zip](https://github.com/Azure/service-fabric-issues/files/1081774/fabric_traces_5.6.220.9494_131412591192127891_339.zip)
[fabric_traces_5.6.220.9494_131412591192127891_340.zip](https://github.com/Azure/service-fabric-issues/files/1081775/fabric_traces_5.6.220.9494_131412591192127891_340.zip)
[fabric_traces_5.6.220.9494_131412591192127891_338.zip](https://github.com/Azure/service-fabric-issues/files/1081776/fabric_traces_5.6.220.9494_131412591192127891_338.zip)
[fabric_traces_5.6.220.9494_131412591115376320_226.zip](https://github.com/Azure/service-fabric-issues/files/1081778/fabric_traces_5.6.220.9494_131412591115376320_226.zip)
[fabric_traces_5.6.220.9494_131412591115376320_225.zip](https://github.com/Azure/service-fabric-issues/files/1081777/fabric_traces_5.6.220.9494_131412591115376320_225.zip)
[fabric_traces_5.6.220.9494_131412591150708220_444.zip](https://github.com/Azure/service-fabric-issues/files/1081780/fabric_traces_5.6.220.9494_131412591150708220_444.zip)
[fabric_traces_5.6.220.9494_131412591150708220_443.zip](https://github.com/Azure/service-fabric-issues/files/1081781/fabric_traces_5.6.220.9494_131412591150708220_443.zip)

**Comment by prinupmathew** [comment]
@suchiagicha 

Other set of logs after initializing ActorProxyFactory  in ScheduleService constructor

[fabric_traces_5.5.216.0_131406550192469332_347.zip](https://github.com/Azure/service-fabric-issues/files/1081923/fabric_traces_5.5.216.0_131406550192469332_347.zip)
[fabric_traces_5.6.220.9494_131412591192127891_342.zip](https://github.com/Azure/service-fabric-issues/files/1081922/fabric_traces_5.6.220.9494_131412591192127891_342.zip)
[fabric_traces_5.6.220.9494_131412591115376320_228.zip](https://github.com/Azure/service-fabric-issues/files/1081929/fabric_traces_5.6.220.9494_131412591115376320_228.zip)
[fabric_traces_5.6.220.9494_131412591115376320_227.zip](https://github.com/Azure/service-fabric-issues/files/1081928/fabric_traces_5.6.220.9494_131412591115376320_227.zip)
[fabric_traces_5.6.220.9494_131412591150708220_447.zip](https://github.com/Azure/service-fabric-issues/files/1081932/fabric_traces_5.6.220.9494_131412591150708220_447.zip)
[fabric_traces_5.6.220.9494_131412591150708220_446.zip](https://github.com/Azure/service-fabric-issues/files/1081931/fabric_traces_5.6.220.9494_131412591150708220_446.zip)

**Comment by prinupmathew** [comment]
@suchiagicha,

I have tried with unsecure cluster setup, it's stuck in the same place where the actor client trying to communicate to actor service.

I am kind of stuck where i can't move forward with POC i am developing for my client.

Any help to identify the root cause is greatly appreciated

**Comment by prinupmathew** [comment]
@suchiagicha,

Attached the ETL logs from AWS unsecure cluster nodes (3 nodes). Please let me know if you find any area that is causing delay between  actor client to actor service.

Here the possible suspect for causing delay is the actor service ScheduleWorkerActor

[fabric_traces_5.6.220.9494_131423560989523731_70.zip](https://github.com/Azure/service-fabric-issues/files/1088620/fabric_traces_5.6.220.9494_131423560989523731_70.zip)
[fabric_traces_5.6.220.9494_131423560989523731_71.zip](https://github.com/Azure/service-fabric-issues/files/1088619/fabric_traces_5.6.220.9494_131423560989523731_71.zip)
[fabric_traces_5.6.220.9494_131423560989523731_68.zip](https://github.com/Azure/service-fabric-issues/files/1088617/fabric_traces_5.6.220.9494_131423560989523731_68.zip)
[fabric_traces_5.6.220.9494_131423560989523731_69.zip](https://github.com/Azure/service-fabric-issues/files/1088618/fabric_traces_5.6.220.9494_131423560989523731_69.zip)
[fabric_traces_5.6.220.9494_131423560468227626_81.zip](https://github.com/Azure/service-fabric-issues/files/1088621/fabric_traces_5.6.220.9494_131423560468227626_81.zip)
[fabric_traces_5.6.220.9494_131423560468227626_82.zip](https://github.com/Azure/service-fabric-issues/files/1088622/fabric_traces_5.6.220.9494_131423560468227626_82.zip)
[fabric_traces_5.6.220.9494_131423560468227626_79.zip](https://github.com/Azure/service-fabric-issues/files/1088623/fabric_traces_5.6.220.9494_131423560468227626_79.zip)
[fabric_traces_5.6.220.9494_131423560468227626_80.zip](https://github.com/Azure/service-fabric-issues/files/1088624/fabric_traces_5.6.220.9494_131423560468227626_80.zip)
[fabric_traces_5.6.220.9494_131423560973756776_25.zip](https://github.com/Azure/service-fabric-issues/files/1088626/fabric_traces_5.6.220.9494_131423560973756776_25.zip)
[fabric_traces_5.6.220.9494_131423560973756776_24.zip](https://github.com/Azure/service-fabric-issues/files/1088627/fabric_traces_5.6.220.9494_131423560973756776_24.zip)

**Comment by suchiagicha** [comment]
@prinupmathew

Your previous run is consuming time in certificate validation in those machines. For the secure cluster , you might see high latency for first request for those machines.  As suggested , you should cache proxyFactory for each client process. Then this issue will be seen once for each process.  .You could also prime the first request in client process, so that your gateway client wont see the high latency .

For second unsecure run ,  looks like you are still creating proxyFactory for each iteration/Call.

Read the section about Service Proxy Lifetime for other questions.
https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-remoting

Here is the counters link
https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-actors-diagnostics

**Comment by VipulM-MSFT** [comment]
@prinupmathew - We hope that you have been able to move forward with your POC. Could we close this issue? If there are other questions or issues, feel free to open another issue.

**Comment by prinupmathew** [comment]
Yes you can close the issue
